### ISSUE TITLE
TensorFlow: lack of repeatability in tf.estimator.DNNClassifier

### ISSUE THREAD

**Issue by IvanDeFalco** [tensorflow/tensorflow#13744]
Please go to Stack Overflow for help and support:

https://stackoverflow.com/questions/tagged/tensorflow

If you open a GitHub issue, here is our policy:

1. It must be a bug or a feature request.
2. The form below must be filled out.
3. It shouldn't be a TensorBoard issue. Those go [here](https://github.com/tensorflow/tensorboard/issues).

**Here's why we have that policy**: TensorFlow developers respond to issues. We want to focus on work that benefits the whole community, e.g., fixing bugs and adding features. Support only helps individuals. GitHub also notifies thousands of people when issues are filed. We want them to see you communicating an interesting problem, rather than being redirected to Stack Overflow.

------------------------

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
- **TensorFlow installed from (source or binary)**:
- **TensorFlow version (use command below)**:
- **Python version**: 
- **Bazel version (if compiling from source)**:
- **CUDA/cuDNN version**:
- **GPU model and memory**:
- **Exact command to reproduce**:

custom code: no, it is the one in https://www.tensorflow.org/get_started/estimator
system: Apple
OS: Mac OsX 10.13
TensorFlow version: 1.3.0
Python version: 3.6.3
GPU model: AMD FirePro D700 (actually, two such GPUs)

You can collect some of this information using our environment capture script:

https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh

You can obtain the TensorFlow version with

python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"

### Describe the problem
Describe the problem clearly here. Be sure to convey here why it's a bug in TensorFlow or a feature request.

Dear all,

I wish to use TensorFlow to perform classification on medical datasets.
To do this, I am using exactly the same code as that proposed in https://www.tensorflow.org/get_started/estimator, the only difference being that I face medical datasets rather than the Iris database. So, it is not custom code.

I do need help about the following problem: if I run the same code on the same data with the same parameters for the network configuration (number of layers, number of neurons in each layer, and so on) , the results are different. I mean, I run the same code ten times, and I obtain ten different values for the accuracy. These values are even largely different, as they range from 73% up to 83%.
This means that the subjects considered suffering from a given disease vary from a run to another. Differently said, once set a network structure, there are several subjects who are considered either healthy or sick depending on the run only.
As you can imagine, this lack of repeatability makes that code useless from both scientific and medical viewpoints: another user, running the same configuration over the same data set, would find a different model and different results, so would cure different subjects. 
I have noticed that for the Iris database the problem seems not to take place, and accuracy is always 0.9666. This depends on the problem being very easy (linearly separable for all but one items, and very small data set). 

I have carried out a search on the internet, and I have found several other people who have noted the same problem. As for the possible solutions, I have read several as well, I have implemented them all, with no result.

Here I add a short list of some suggested remedies that failed in my case:

os.environ['PYTHONHASHSEED'] = '0'
np.random.seed(0)
tf.set_random_seed(0)
rn.seed(0)
tf.reset_default_graph()

session_conf = tf.ConfigProto(
    intra_op_parallelism_threads=1,
    inter_op_parallelism_threads=1
)
sess = tf.Session(graph=tf.get_default_graph(), config=session_conf)

Is there any chance to fix this problem? It is a pity that such an excellent tool, as TensorFlow is, cannot guarantee repeatibility.

I am using TensorFlow 1.3.0, Python 3.6.3, and an Apple with Mac OsX 10.13.

Thank you very much!

Best regards

Ivanoe De Falco

### Source code / logs
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached. Try to provide a reproducible test case that is the bare minimum necessary to generate the problem.

Source code is in:
https://www.tensorflow.org/get_started/estimator

**Comment by facaiy** [comment]
Do you shuffle input data?

**Comment by IvanDeFalco** [comment]
Dear facaiy,
thank you very much for your fast reply!
Actually, yesterday evening I noticed the same issue, so I set shuffle=False, and performed several tests.
Now all the runs yield the same result, so your idea is good and very helpful for my goal, thank you!
Actually, generally speaking, setting shuffle=False is just a way to get around the real problem, because the problem still exists: even if one sets all the random seeds fixed, the shuffling still produces different results. 
I feel there should be an easy way to declare whether or not a DNN run should be based on a fixed seed so as to make it repeatable. This would be, I believe, a great improvement in the ease of use of TensorFlow, and I hope you good guys will work on it.
By now, thanks a lot for your cooperation!
Ivan

**Comment by facaiy** [comment]
Hi, @IvanDeFalco . If I understand you correctly, you want a global seed setter for random, like `numpy.random.seed(0)`, right?

Sounds reasonable, however, it goes beyond my scope and perhaps difficult to implement. Let's wait for reply from official tensorflowers.

**Comment by IvanDeFalco** [comment]
yes @facaiy it is exactly what I am looking for.
Let's hope somebody will start working on this.
Again, thanks :)

**Comment by facaiy** [comment]
Perhaps it is nearly impossible for multiple thread or distributed environment.

**Comment by ekelsen** [comment]
Just to double check - you're running this on the CPU? Or using OpenCL on the AMD GPUs?

**Comment by IvanDeFalco** [comment]
Dear  **@ekelsen** Currently the program is running on the CPU.

**Comment by skye** [comment]
@mrry do you know if there's a way to deterministically seed shuffled input data?

**Comment by mrry** [comment]
@skyewm: Yes, the `Dataset.shuffle()` transformation will pick up the graph seed, if one is specified (or you can pass an explicit seed as an optional argument).

**Comment by aselle** [comment]
@IvanDeFalco, does this suggestion resolve your issue.

**Comment by mrry** [comment]
Closing this due to lack of activity.

**Comment by secsilm** [comment]
Without `tf.set_random_seed`, running multiple training in the same `model_dir` cause a fake accuracy improvement. But with `tf.set_random_seed`, it's fine. Why is this related with `tf.set_random_seed`?

Here is the question I asked on SO: [python - Weird accuracy improvement by multiple small epochs training - Stack Overflow](https://stackoverflow.com/questions/51629819/weird-accuracy-improvement-by-multiple-small-epochs-training)

Thanks.
